src/link-preview-expander.ts

```
@@ -1,16 +1,142 @@
+            provideHover: (document, position) => {
+                const range = document.getWordRangeAtPosition(position)
+                if (!range) {
+                    return null
                 }
-            }),
+                const maybeURL = cleanURL(getWord(document, range))
+                const isURL = checkIsURL(maybeURL)
+
+                if (!isURL) {
+                    return null
+                }
+
+                /**
+                 * Creates the markdown string to be rendered in the hover tooltip.
+                 */
+                const createResult: (metadata?: Metadata) => sourcegraph.Badged<sourcegraph.Hover> = metadata => {
+                    const { image, title, description } = metadata ?? {}
+
+                    let markdownContent = `#### [${title || maybeURL}](${maybeURL})\n\n`
+
+                    if (image || description) {
+                        markdownContent += '---\n\n'
+
+                        if (image) {
+                            markdownContent += `<img height="64" src="${image}" align="left" />`
+                        }
+
+                        if (description) {
+                            markdownContent += `<p>${description ?? ''}</p>`
+                        }
+                    }
+
+                    return {
+                        contents: {
+                            value: markdownContent,
+                            kind: sourcegraph.MarkupKind.Markdown,
+                        },
+                    }
+                }
+
+                // TODO(tj): Return async iterable (once allowed) instead of promise so that we can show link before metadata loads
+                return fetch('https://cors-anywhere.herokuapp.com/' + maybeURL, { cache: 'force-cache' })
+                    .then(response => response.text())
+                    .then(text => createResult(getMetadataFromHTMLString(text)))
+                    .catch(() => createResult())
```

Any reason to not let errors bubble to the caller?

---

To render links regardless of error. Fetch request could be flaky, or the user could be authenticated to a link that returns 401 (e.g. Figma links).